File: Tests/Strata.Foundation.CoreTest/test/java/strata/foundation/core/utility/ExpendableTest.java
/// ///////////////////////////////////////////////////////////////////////////
// LimitedTest.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.NoSuchElementException;

import static org.junit.jupiter.api.Assertions.*;

@Tag("CommitStage")
public
class ExpendableTest
{
    @Test
    public void
    testConstructors()
    {
        Expendable<String> subject1 = new Expendable<>();
        Expendable<String> subject2 = new Expendable<>("subject2");
        Expendable<String> subject3 = new Expendable<>("subject3",5);

        assertEquals(0, subject1.getAllowed());
        assertEquals(0,subject1.getRemaining());

        try
        {
            subject1.get();
            fail("Expected an exception to be thrown");
        }
        catch (Exception e) {}

        assertEquals(1, subject2.getAllowed());
        assertEquals(1,subject2.getRemaining());
        assertEquals("subject2",subject2.get());

        assertEquals(5, subject3.getAllowed());

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertEquals(5-i, subject3.getRemaining());
            assertEquals("subject3",subject3.get());
        }

        assertEquals(0, subject3.getRemaining());

        try
        {
            subject3.get();
            fail("Expected an exception to be thrown");
        }
        catch (Exception e) {}

    }

    @Test
    public void
    testEmptyAndOf()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertEquals(0, subject1.getAllowed());
        assertEquals(0,subject1.getRemaining());

        try
        {
            subject1.get();
            fail("Expected an exception to be thrown");
        }
        catch (Exception e) {}

        assertEquals(1, subject2.getAllowed());
        assertEquals(1,subject2.getRemaining());
        assertEquals("subject2",subject2.get());

        assertEquals(5, subject3.getAllowed());

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertEquals(5-i, subject3.getRemaining());
            assertEquals("subject3",subject3.get());
        }

        assertEquals(0, subject3.getRemaining());

        try
        {
            subject3.get();
            fail("Expected an exception to be thrown");
        }
        catch (Exception e) {}

    }

    @Test
    public void
    testIfPresent()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertFalse(subject1.isPresent());
        assertTrue(subject1.isEmpty());
        subject1.ifPresent(value -> fail("Expected subject1 to be empty"));

        assertTrue(subject2.isPresent());
        assertFalse(subject2.isEmpty());
        subject2.ifPresent(value -> assertEquals("subject2", value));
        assertFalse(subject2.isPresent());
        assertTrue(subject2.isEmpty());
        subject2.ifPresent(value -> fail("Expected subject2 to be empty"));

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertTrue(subject3.isPresent());
            assertFalse(subject3.isEmpty());
            subject3.ifPresent(value -> assertEquals("subject3", value));
        }

        assertFalse(subject3.isPresent());
        assertTrue(subject3.isEmpty());
        subject3.ifPresent(value -> fail("Expected subject3 to be empty"));

    }

    @Test
    public void
    testIfPresentOrElse()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertFalse(subject1.isPresent());
        assertTrue(subject1.isEmpty());
        assertEquals(
            "subject 1 is empty",
            subject1.ifPresentOrElse(value -> value, () -> "subject 1 is empty"));

        assertTrue(subject2.isPresent());
        assertFalse(subject2.isEmpty());
        assertEquals(
            "subject2",
            subject2.ifPresentOrElse(value -> value, () -> "subject 2 is empty"));
        assertFalse(subject2.isPresent());
        assertTrue(subject2.isEmpty());
       assertEquals(
            "subject 2 is empty",
            subject2.ifPresentOrElse(value -> value, () -> "subject 2 is empty"));

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertTrue(subject3.isPresent());
            assertFalse(subject3.isEmpty());
            assertEquals(
                "subject3",
                subject3.ifPresentOrElse(value -> value, () -> "subject 3 is empty"));
        }

        assertFalse(subject3.isPresent());
        assertTrue(subject3.isEmpty());
        assertEquals(
            "subject 3 is empty",
            subject3.ifPresentOrElse(value -> value, () -> "subject 3 is empty"));

    }

    @Test
    public void
    testIfPresentOrThrow()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertFalse(subject1.isPresent());
        assertTrue(subject1.isEmpty());
        assertThrows(
            NoSuchElementException.class,
            () -> subject1.ifPresentOrThrow(
                value -> value,
                new NoSuchElementException("subject 1 is empty")));

        assertTrue(subject2.isPresent());
        assertFalse(subject2.isEmpty());
        assertEquals(
            "subject2",
            subject2.ifPresentOrThrow(
                value -> value,
                new NoSuchElementException("subject 2 is empty")));

        assertFalse(subject2.isPresent());
        assertTrue(subject2.isEmpty());
        assertThrows(
            NoSuchElementException.class,
            () -> subject2.ifPresentOrThrow(
                value -> value,
                new NoSuchElementException("subject 2 is empty")));

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertTrue(subject3.isPresent());
            assertFalse(subject3.isEmpty());
            assertEquals(
                "subject3",
                subject3.ifPresentOrThrow(
                    value -> value,
                    new NoSuchElementException("subject 3 is empty")));
        }

        assertFalse(subject3.isPresent());
        assertTrue(subject3.isEmpty());
        assertThrows(
            NoSuchElementException.class,
            () -> subject3.ifPresentOrThrow(
                value -> value,
                new NoSuchElementException("subject 3 is empty")));

    }

    @Test
    public void
    testIfPresentOrElseNoReturn()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertFalse(subject1.isPresent());
        assertTrue(subject1.isEmpty());
        subject1.ifPresentOrElseNoReturn(
            value -> fail("subject 1 should be empty"),
            () -> System.out.println("subject 1 is empty"));

        assertTrue(subject2.isPresent());
        assertFalse(subject2.isEmpty());
        subject2.ifPresentOrElseNoReturn(
            value -> assertEquals("subject2",value),
            () -> {throw new IllegalStateException("subject 2 is empty");});
        assertFalse(subject2.isPresent());
        assertTrue(subject2.isEmpty());
        subject2.ifPresentOrElseNoReturn(
            value -> fail("subject 2 is empty"),
            () -> System.out.println("subject 2 is empty"));

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertTrue(subject3.isPresent());
            assertFalse(subject3.isEmpty());
            subject3.ifPresentOrElseNoReturn(
                value -> assertEquals("subject3",value),
                () -> {throw new IllegalStateException("subject 3 is empty");});
        }

        assertFalse(subject3.isPresent());
        assertTrue(subject3.isEmpty());
        subject3.ifPresentOrElseNoReturn(
            value -> fail("subject 3 is empty"),
            () -> System.out.println("subject 3 is empty"));

    }

    @Test
    public void
    testIfPresentOrThrowNoReturn()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertFalse(subject1.isPresent());
        assertTrue(subject1.isEmpty());
        assertThrows(
            NoSuchElementException.class,
            () -> subject1.ifPresentOrThrowNoReturn(
                value -> fail("subject 1 should be empty"),
                new NoSuchElementException("subject 1 is empty")));

        assertTrue(subject2.isPresent());
        assertFalse(subject2.isEmpty());
        subject2.ifPresentOrThrowNoReturn(
            value -> assertEquals("subject2",value),
            new IllegalStateException("subject 2 is empty"));

        assertFalse(subject2.isPresent());
        assertTrue(subject2.isEmpty());
        assertThrows(
            NoSuchElementException.class,
            () -> subject2.ifPresentOrThrowNoReturn(
                value -> fail("subject 2 is empty"),
                new NoSuchElementException("subject 2 is empty")));

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertTrue(subject3.isPresent());
            assertFalse(subject3.isEmpty());
            subject3.ifPresentOrThrowNoReturn(
                value -> assertEquals("subject3",value),
                new IllegalStateException("subject 3 is empty"));
        }

        assertFalse(subject3.isPresent());
        assertTrue(subject3.isEmpty());
        assertThrows(
            NoSuchElementException.class,
            () -> subject3.ifPresentOrThrowNoReturn(
                value -> fail("subject 3 is empty"),
                new NoSuchElementException("subject 3 is empty")));
    }

    @Test
    public void
    testFilter()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertTrue(subject1.filter(value -> value.equals("subject1")).isEmpty());
        assertTrue(subject2.filter(value -> value.equals("subject1")).isEmpty());
        assertTrue(subject2.filter(value -> value.equals("subject2")).isPresent());
        assertEquals("subject2", subject2.filter(value -> value.equals("subject2")).get());

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertTrue(subject3.filter(value -> value.equals("subject1")).isEmpty());
            assertTrue(subject3.filter(value -> value.equals("subject3")).isPresent());
            assertEquals("subject3", subject3.filter(value -> value.equals("subject3")).get());
        }
    }

    @Test
    public void
    testMap()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertTrue(subject1.map(value -> value.toUpperCase()).isEmpty());
        assertEquals(
            "SUBJECT2+extra",
            subject2
                .map(value -> value.toUpperCase())
                .map(value -> value + "+extra").get());
        assertTrue(subject2.isEmpty());


        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertEquals("SUBJECT3", subject3.map(value -> value.toUpperCase()).get());
        }

        assertTrue(subject3.isEmpty());

    }

    @Test
    public void
    testFlatMap()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertTrue(subject1.flatMap(value -> Expendable.of(value.toUpperCase())).isEmpty());
        assertEquals(
            "SUBJECT2+extra",
            subject2
                .flatMap(value -> Expendable.of(value.toUpperCase()))
                .flatMap(value -> Expendable.of(value + "+extra"))
                .get());
        assertTrue(subject2.isEmpty());

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertEquals("SUBJECT3", subject3.flatMap(value -> Expendable.of(value.toUpperCase())).get());
        }

        assertTrue(subject3.isEmpty());
    }

    @Test
    public void
    testOrElse()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertEquals("default", subject1.orElse("default"));
        assertEquals("subject2", subject2.orElse("default"));
        assertEquals("default", subject2.orElse("default"));

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertEquals("subject3", subject3.orElse("default"));
        }

        assertEquals("default", subject3.orElse("default"));

    }

    @Test
    public void
    testOrElseGet()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertEquals("default", subject1.orElseGet(() -> "default"));
        assertEquals("subject2", subject2.orElseGet(() -> "default"));
        assertEquals("default", subject2.orElseGet(() -> "default"));

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertEquals("subject3", subject3.orElseGet(() -> "default"));
        }

        assertEquals("default", subject3.orElseGet(() -> "default"));

    }

    @Test
    public void
    testOrElseThrow()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertThrows(
            NoSuchElementException.class,
            () -> subject1.orElseThrow(
                () -> new NoSuchElementException("subject 1 is empty")));

        assertEquals("subject2", subject2.orElseThrow(
            () -> new NoSuchElementException("subject 2 is empty")));

        assertThrows(
            NoSuchElementException.class,
            () -> subject2.orElseThrow(
                () -> new NoSuchElementException("subject 2 is empty")));

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertEquals("subject3", subject3.orElseThrow(
                () -> new NoSuchElementException("subject 3 is empty")));
        }

        assertThrows(
            NoSuchElementException.class,
            () -> subject3.orElseThrow(
                () -> new NoSuchElementException("subject 3 is empty")));
    }

    @Test
    public void
    testIsExpended()
    {
        Expendable<String> subject1 = Expendable.empty();
        Expendable<String> subject2 = Expendable.of("subject2");
        Expendable<String> subject3 = Expendable.of("subject3",5);

        assertTrue(subject1.isExpended());
        assertFalse(subject2.isExpended());
        subject2.get();
        assertTrue(subject2.isExpended());

        for (int i = 0; i < 5; i++)
        {
            System.out.println("i = " + i);
            assertFalse(subject3.isExpended());
            subject3.get();
        }

        assertTrue(subject3.isExpended());
    }
}

//////////////////////////////////////////////////////////////////////////////

File: Tests/Strata.Foundation.CoreTest/test/typescript/Utility/ExpendableTest.ts
import "jest"
import {Expendable} from "strata.foundation.core/Utility";
import {NoSuchElementException} from "strata.foundation.core/Utility";

describe(
    "Expendable",
    () =>
    {
        it(
            "testConstructors",
            () =>
            {
                const subject1: Expendable<string> = new Expendable<string>();
                const subject2: Expendable<string> = new Expendable<string>("subject2");
                const subject3: Expendable<string> = new Expendable<string>("subject3", 5);

                expect(subject1.getAllowed()).toBe(0);
                expect(subject1.getRemaining()).toBe(0);
                expect(subject1.isPresent()).toBeFalsy();
                expect(subject1.isEmpty()).toBeTruthy();

                try
                {
                    subject1.get();
                    fail("should have thrown exception");
                }
                catch (e)
                {
                    expect(e instanceof NoSuchElementException).toBeTruthy();
                }

                expect(subject2.getAllowed()).toBe(1);
                expect(subject2.getRemaining()).toBe(1);
                expect(subject2.isPresent()).toBeTruthy();
                expect(subject2.get()).toBe("subject2");
                expect(subject2.getRemaining()).toBe(0);
                expect(subject2.isPresent()).toBeFalsy();
                expect(subject2.isEmpty()).toBeTruthy();


            });
        it(
            "testOfAndEmpty",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.getAllowed()).toBe(0);
                expect(subject1.getRemaining()).toBe(0);
                expect(subject1.isPresent()).toBeFalsy();
                expect(subject1.isEmpty()).toBeTruthy();

                try
                {
                    subject1.get();
                    fail("should have thrown exception");
                }
                catch (e)
                {
                    expect(e instanceof NoSuchElementException).toBeTruthy();
                }

                expect(subject2.getAllowed()).toBe(1);
                expect(subject2.getRemaining()).toBe(1);
                expect(subject2.isPresent()).toBeTruthy();
                expect(subject2.get()).toBe("subject2");
                expect(subject2.getRemaining()).toBe(0);
                expect(subject2.isPresent()).toBeFalsy();
                expect(subject2.isEmpty()).toBeTruthy();
            });
        it(
            "testIfPresent",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.getAllowed()).toBe(0);
                expect(subject1.getRemaining()).toBe(0);
                expect(subject1.isPresent()).toBeFalsy();
                expect(subject1.isEmpty()).toBeTruthy();
                subject1.ifPresent(actual => fail("expected subject 1 to be empty"));

                expect(subject2.getAllowed()).toBe(1);
                expect(subject2.getRemaining()).toBe(1);
                expect(subject2.isPresent()).toBeTruthy();
                subject2.ifPresent(actual => expect(actual).toBe("subject2"));
                expect(subject2.getRemaining()).toBe(0);
                expect(subject2.isPresent()).toBeFalsy();
                expect(subject2.isEmpty()).toBeTruthy();
                subject2.ifPresent(actual => fail("expected subject 2 to be empty"));

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.isPresent()).toBeTruthy();
                    expect(subject3.isEmpty()).toBeFalsy();
                    expect(subject3.getRemaining()).toBe(5 - i);
                    subject3.ifPresent(actual => expect(actual).toBe("subject3"));
                }

                expect(subject3.isPresent()).toBeFalsy();
                expect(subject3.isEmpty()).toBeTruthy();
                expect(subject3.getRemaining()).toBe(0);
                subject3.ifPresent(actual => fail("expected subject 3 to be empty"));
            });
        it(
            "testIfPresentOrElse",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.getAllowed()).toBe(0);
                expect(subject1.getRemaining()).toBe(0);
                expect(subject1.isPresent()).toBeFalsy();
                expect(subject1.isEmpty()).toBeTruthy();
                expect(
                    subject1.ifPresentOrElse(
                        value => value,
                        () => "subject1 is empty"))
                    .toBe("subject1 is empty");

                expect(subject2.getAllowed()).toBe(1);
                expect(subject2.getRemaining()).toBe(1);
                expect(subject2.isPresent()).toBeTruthy();
                expect(
                    subject2.ifPresentOrElse(
                        actual => actual,
                        () => "subject2 is empty"))
                    .toBe("subject2");
                expect(subject2.getRemaining()).toBe(0);
                expect(subject2.isPresent()).toBeFalsy();
                expect(subject2.isEmpty()).toBeTruthy();
                expect(
                    subject2.ifPresentOrElse(
                        actual => actual,
                        () => "subject2 is empty"))
                    .toBe("subject2 is empty");

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.isPresent()).toBeTruthy();
                    expect(subject3.isEmpty()).toBeFalsy();
                    expect(subject3.getRemaining()).toBe(5 - i);
                    expect(
                        subject3.ifPresentOrElse(
                            actual => actual,
                            () => "subject3 is empty"))
                        .toBe("subject3");
                }

                expect(subject3.isPresent()).toBeFalsy();
                expect(subject3.isEmpty()).toBeTruthy();
                expect(subject3.getRemaining()).toBe(0);
                expect(
                    subject3.ifPresentOrElse(
                        actual => actual,
                        () => "subject3 is empty"))
                    .toBe("subject3 is empty");
            });
        it(
            "testIfPresentOrThrow",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.getAllowed()).toBe(0);
                expect(subject1.getRemaining()).toBe(0);
                expect(subject1.isPresent()).toBeFalsy();
                expect(subject1.isEmpty()).toBeTruthy();
                expect(
                    () => subject1.ifPresentOrThrow(
                        value => value,
                        new NoSuchElementException("subject1 is empty"))
                    .toThrow(new NoSuchElementException("subject1 is empty")));

                expect(subject2.getAllowed()).toBe(1);
                expect(subject2.getRemaining()).toBe(1);
                expect(subject2.isPresent()).toBeTruthy();
                expect(
                    subject2.ifPresentOrThrow(
                        actual => actual,
                        new NoSuchElementException("subject2 is empty")))
                    .toBe("subject2");
                expect(subject2.getRemaining()).toBe(0);
                expect(subject2.isPresent()).toBeFalsy();
                expect(subject2.isEmpty()).toBeTruthy();
                expect(
                    () => subject2.ifPresentOrThrow(
                        actual => actual,
                        new NoSuchElementException("subject2 is empty")))
                    .toThrow(new NoSuchElementException("subject2 is empty"));

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.isPresent()).toBeTruthy();
                    expect(subject3.isEmpty()).toBeFalsy();
                    expect(subject3.getRemaining()).toBe(5 - i);
                    expect(
                        subject3.ifPresentOrThrow(
                            actual => actual,
                            new NoSuchElementException("subject3 is empty")))
                        .toBe("subject3");
                }

                expect(subject3.isPresent()).toBeFalsy();
                expect(subject3.isEmpty()).toBeTruthy();
                expect(subject3.getRemaining()).toBe(0);
                expect(
                    () => subject3.ifPresentOrThrow(
                        actual => actual,
                        new NoSuchElementException("subject3 is empty")))
                    .toThrow(new NoSuchElementException("subject3 is empty"));
            });
        it(
            "testIfPresentOrElseNoReturn",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.getAllowed()).toBe(0);
                expect(subject1.getRemaining()).toBe(0);
                expect(subject1.isPresent()).toBeFalsy();
                expect(subject1.isEmpty()).toBeTruthy();
                subject1.ifPresentOrElseNoReturn(
                        value => fail("expected subject 1 to be empty"),
                        () => console.log("subject1 is empty"));

                expect(subject2.getAllowed()).toBe(1);
                expect(subject2.getRemaining()).toBe(1);
                expect(subject2.isPresent()).toBeTruthy();
                subject2.ifPresentOrElseNoReturn(
                    actual => expect(actual).toBe("subject2"),
                    () => fail("subject2 is empty"));

                expect(subject2.getRemaining()).toBe(0);
                expect(subject2.isPresent()).toBeFalsy();
                expect(subject2.isEmpty()).toBeTruthy();
                subject2.ifPresentOrElseNoReturn(
                    actual => fail("subject2 should be empty"),
                    () => console.log("subject2 is empty"));

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.isPresent()).toBeTruthy();
                    expect(subject3.isEmpty()).toBeFalsy();
                    expect(subject3.getRemaining()).toBe(5 - i);
                    subject3.ifPresentOrElseNoReturn(
                        actual => expect(actual).toBe("subject3"),
                        () => fail("subject3 is empty"));
                }

                expect(subject3.isPresent()).toBeFalsy();
                expect(subject3.isEmpty()).toBeTruthy();
                expect(subject3.getRemaining()).toBe(0);
                subject3.ifPresentOrElseNoReturn(
                    actual => fail("subject3 should be empty"),
                    () => console.log("subject3 is empty"));
            });
        it(
            "testIfPresentOrThrowNoReturn",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.getAllowed()).toBe(0);
                expect(subject1.getRemaining()).toBe(0);
                expect(subject1.isPresent()).toBeFalsy();
                expect(subject1.isEmpty()).toBeTruthy();
                expect(
                    () => subject1.ifPresentOrThrowNoReturn(
                        value => value,
                        new NoSuchElementException("subject1 is empty")))
                    .toThrow(new NoSuchElementException("subject1 is empty"));

                expect(subject2.getAllowed()).toBe(1);
                expect(subject2.getRemaining()).toBe(1);
                expect(subject2.isPresent()).toBeTruthy();
                subject2.ifPresentOrThrowNoReturn(
                    actual => expect(actual).toBe("subject2"),
                    new NoSuchElementException("subject2 is empty"));
                expect(subject2.getRemaining()).toBe(0);
                expect(subject2.isPresent()).toBeFalsy();
                expect(subject2.isEmpty()).toBeTruthy();
                expect(
                    () => subject2.ifPresentOrThrowNoReturn(
                        actual => actual,
                        new NoSuchElementException("subject2 is empty")))
                    .toThrow(new NoSuchElementException("subject2 is empty"));

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.isPresent()).toBeTruthy();
                    expect(subject3.isEmpty()).toBeFalsy();
                    expect(subject3.getRemaining()).toBe(5 - i);
                    subject3.ifPresentOrThrowNoReturn(
                        actual => expect(actual).toBe("subject3"),
                        new NoSuchElementException("subject3 is empty"));
                }

                expect(subject3.isPresent()).toBeFalsy();
                expect(subject3.isEmpty()).toBeTruthy();
                expect(subject3.getRemaining()).toBe(0);
                expect(
                    () => subject3.ifPresentOrThrowNoReturn(
                        actual => actual,
                        new NoSuchElementException("subject3 is empty")))
                    .toThrow(new NoSuchElementException("subject3 is empty"));
            });
        it(
            "testFilter",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.filter(value => value === "subject1").isEmpty()).toBeTruthy();
                expect(subject2.filter(value => value === "subject1").isEmpty()).toBeTruthy();
                expect(subject2.filter(value => value === "subject2").isPresent()).toBeTruthy();
                expect(subject2.filter(value => value === "subject2").get()).toBe("subject2");

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.filter(value => value === "subject1").isEmpty()).toBeTruthy();
                    expect(subject3.filter(value => value === "subject3").isPresent()).toBeTruthy();
                    expect(subject3.filter(value => value === "subject3").get()).toBe("subject3");
                }

            });

        it(
            "testMap",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.map(value => value.toUpperCase()).isEmpty()).toBeTruthy();
                expect(subject2.map(value => value.toUpperCase()).get()).toBe("SUBJECT2");

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.map(value => value.toUpperCase()).get()).toBe("SUBJECT3");
                }

                expect(subject3.isEmpty()).toBeTruthy();
            });
        it(
            "testFlatMap",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.flatMap(value => Expendable.of(value.toUpperCase())).isEmpty()).toBeTruthy();
                expect(subject2.flatMap(value => Expendable.of(value.toUpperCase())).get()).toBe("SUBJECT2");
                expect(subject2.isEmpty()).toBeTruthy();

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.flatMap(value => Expendable.of(value.toUpperCase())).get()).toBe("SUBJECT3");
                }

                expect(subject3.isEmpty()).toBeTruthy();
            });
        it(
            "testOrElse",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.orElse("default")).toBe("default");
                expect(subject2.orElse("default")).toBe("subject2");
                expect(subject2.orElse("default")).toBe("default");

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.orElse("default")).toBe("subject3");
                }

                expect(subject3.orElse("default")).toBe("default");
            });
        it(
            "testOrElseGet",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.orElseGet(() => "default")).toBe("default");
                expect(subject2.orElseGet(() => "default")).toBe("subject2");
                expect(subject2.orElseGet(() => "default")).toBe("default");

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.orElseGet(() => "default")).toBe("subject3");
                }

                expect(subject3.orElseGet(() => "default")).toBe("default");
            });
        it(
            "testOrElseThrow",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(() => subject1.orElseThrow(new NoSuchElementException("subject1 is empty")))
                    .toThrow(new NoSuchElementException("subject1 is empty"));
                expect(subject2.orElseThrow(new NoSuchElementException("subject2 is empty"))).toBe("subject2");
                expect(() => subject2.orElseThrow(new NoSuchElementException("subject2 is empty")))
                    .toThrow(new NoSuchElementException("subject2 is empty"));

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.orElseThrow(new NoSuchElementException("subject3 is empty"))).toBe("subject3");
                }

                expect(() => subject3.orElseThrow(new NoSuchElementException("subject3 is empty")))
                    .toThrow(new NoSuchElementException("subject3 is empty"));
            });
        it(
            "isExpended",
            () =>
            {
                const subject1: Expendable<string> = Expendable.empty();
                const subject2: Expendable<string> = Expendable.of("subject2");
                const subject3: Expendable<string> = Expendable.of("subject3", 5);

                expect(subject1.isExpended()).toBeTruthy();
                expect(subject2.isExpended()).toBeFalsy();

                subject2.get();
                expect(subject2.isExpended()).toBeTruthy();

                for (let i: number = 0; i < 5; ++i)
                {
                    console.log("i = " + i);
                    expect(subject3.isExpended()).toBeFalsy();
                    subject3.get();
                }

                expect(subject3.isExpended()).toBeTruthy();
            });
    });